src/utils/query-api.js

```diff
@@ -0,0 +1,146 @@
+ * fetching. These things get their data from external sources like the browser or the
+ * network.
+ *
+ * In most cases you'll want to use RegularExternalCommunicationDelegate.
+ */
+export interface ExternalCommunicationDelegate {
+  // Fetch a cross-origin URL and return its Response. If postData is specified,
+  // the method should be POST.
+  fetchUrlResponse(url: string, postData?: MixedObject): Promise<Response>;
+
+  // Query the symbolication API of the browser, if a connection to the browser
+  // is available.
+  queryBrowserSymbolicationApi(
+    path: string,
+    requestJson: string
+  ): Promise<string>;
+}
+
+export type ApiQueryResult =
+  | { type: 'SUCCESS', responseJson: any }
+  | { type: 'ERROR', errors: ApiQueryError[] };
+
+/**
+ * Sends a JSON query to the browser symbolication API and, if supplied, to a
+ * symbol server.
+ */
+export async function queryApiWithFallback(
+  path: string,
+  requestJson: string,
+  symbolServerUrlForFallback: string | null,
+  delegate: ExternalCommunicationDelegate
+): Promise<ApiQueryResult> {
+  const errors: ApiQueryError[] = [];
+
+  // See if we can get an API result from the browser.
+  try {
+    const response = await delegate.queryBrowserSymbolicationApi(
+      path,
+      requestJson
+    );
+    const responseJson = await JSON.parse(response);
+    if (!responseJson.error) {
+      // The browser gave us a successful response.
+      return { type: 'SUCCESS', responseJson };
+    }
+    errors.push({
+      type: 'BROWSER_API_ERROR',
+      apiErrorMessage: responseJson.error,
+    });
+  } catch (e) {
+    errors.push({
+      type: 'BROWSER_CONNECTION_ERROR',
+      browserConnectionErrorMessage: e.toString(),
+    });
+  }
+
+  // See if we can get sources from a local symbol server.
+  if (symbolServerUrlForFallback !== null) {
+    const url = symbolServerUrlForFallback + path;
+    try {
+      const response = await delegate.fetchUrlResponse(url, requestJson);
+      const responseJson = await response.json();
+      if (!responseJson.error) {
+        // Local symbol server gave us the source. Success!
+        return { type: 'SUCCESS', responseJson };
+      }
+      errors.push({
+        type: 'SYMBOL_SERVER_API_ERROR',
+        apiErrorMessage: responseJson.error,
+      });
+    } catch (e) {
+      errors.push({
+        type: 'NETWORK_ERROR',
+        url,
+        networkErrorMessage: e.toString(),
+      });
+    }
+  }
+
+  return { type: 'ERROR', errors };
+}
+
+export interface ExternalCommunicationCallbacks {
+  onBeginUrlRequest(url: string): void;
+  onBeginBrowserConnectionQuery(): void;
+}
+
+/**
+ * The default implementation of the ExternalCommunicationDelegate interface.
+ * Uses fetch for URL requests and the supplied browser connection for browser
+ * connection requests.
+ *
+ * It also takes an object with callbacks, for loading indicators.
+ */
+export class RegularExternalCommunicationDelegate {
```

Comment:
Nit: it could be nice to annotate this class with `implements ExternalCommunicationDelegate` so we make sure that we keep this class in sync with the interface in the future. Also I find it a bit easir to follow when these things are explicit.
It's optional though because the name already gives away that this implements this interface anyways. So it's up to you if you want to do it.

Reply:
Oh thanks, I forgot `implements` exists.